### PR TITLE
PR: Make index column of DataFrame editor always visible

### LINKED ISSUE #2512: Index column of Pandas DataFrame should be visible all the time
Hello,

Index of Pandas DataFrame should be in a frozen pane in variable explorer

capture d ecran 2015-07-02 a 22 36 20
capture d ecran 2015-07-02 a 22 36 30

Kind regards

### PR DESCRIPTION
**Pull request by andfoy**
Fixes #2512

### CODE REVIEW

**Comment by andfoy** [comment]
@ccordoba12 According to the Qt and the PyQt documentation (4.8), the function ``setResizeMode(ResizeMode mode)`` does exist. However, there are execution errors in all the CI Engines

**Comment by ccordoba12** [comment]
@andfoy, `setResizeMode` is handled differently for Qt4 and 5. Please see this block to see how to do it

https://github.com/spyder-ide/spyder/blob/master/spyder/widgets/projects/explorer.py#L67

**Comment by ccordoba12** [comment]
You also need to rebase your changes on top of latest 3.x to fix the errors in CircleCI

**Comment by andfoy** [comment]
@ccordoba12 All the errors were fixed!

File: spyder/widgets/sourcecode/base.py [hunk 1]
```diff
@@ -867,6 +867,10 @@ def __move_line_or_selection(self, after_current_line=True):
 
         if after_current_line:
             text = to_text_string(cursor.block().text())
+            if len(text) == 0:
+                #If the next line is blank
+                sel_text = sel_text[0:-1]
+                sel_text = os.linesep + sel_text
```
Comment (ccordoba12) [inline spyder/widgets/sourcecode/base.py]: Don't forget to remove these lines and squash its commit

File: spyder/widgets/variableexplorer/dataframeeditor.py [hunk 2]
```diff
@@ -405,6 +407,14 @@ def __init__(self, parent, model):
         QTableView.__init__(self, parent)
         self.setModel(model)
 
+        self.frozenTableView = QTableView(self)
+        self.init_frozen()
+
+        self.horizontalHeader().sectionResized.connect(self.updateSectionWidth)
+        self.verticalHeader().sectionResized.connect(self.updateSectionHeight)
+
+        self.frozenTableView.verticalScrollBar().valueChanged.connect(self.verticalScrollBar().setValue)
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: Please make this fit in 79 columns

File: spyder/widgets/variableexplorer/dataframeeditor.py [hunk 3]
```diff
@@ -415,7 +425,63 @@ def __init__(self, parent, model):
                             lambda val: self.load_more_data(val, columns=True))
         self.verticalScrollBar().valueChanged.connect(
                                lambda val: self.load_more_data(val, rows=True))
+        self.verticalScrollBar().valueChanged.connect(self.frozenTableView.verticalScrollBar().setValue)
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: 79 columns

File: spyder/widgets/variableexplorer/dataframeeditor.py [hunk 4]
```diff
@@ -415,7 +425,63 @@ def __init__(self, parent, model):
...
     
+    def init_frozen(self):
+        self.frozenTableView.setModel(self.model())
+        self.frozenTableView.setFocusPolicy(Qt.NoFocus)
+        self.frozenTableView.verticalHeader().hide()
+        if PYQT5:
+            self.frozenTableView.horizontalHeader().setSectionResizeMode(QHeaderView.Fixed)
+        else:
+            self.frozenTableView.horizontalHeader().setResizeMode(QHeaderView.Fixed)
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: Both lines in this `if` block take more are wider than 79 columns.

File: spyder/widgets/variableexplorer/dataframeeditor.py [hunk 5]
```diff
@@ -415,7 +425,63 @@ def __init__(self, parent, model):
...
+        self.frozenTableView.setRowHeight(logicalIndex, newSize)
+
+    def resizeEvent(self, event):
+        QTableView.resizeEvent(self, event)
+        self.updateFrozenTableGeometry()
+
+    def moveCursor(self, cursorAction, modifiers):
+        current = QTableView.moveCursor(self, cursorAction, modifiers)
+        if cursorAction == self.MoveLeft and current.column() > 1 and self.visualRect(current).topLeft().x() < (self.frozenTableView.columnWidth(0) + self.frozenTableView.columnWidth(1)):
+            newValue = self.horizontalScrollBar().value() + self.visualRect(current).topLeft().x() - (self.frozenTableView.columnWidth(0) + self.frozenTableView.columnWidth(1))
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: 79 columns

File: spyder/widgets/variableexplorer/dataframeeditor.py [hunk 6]
```diff
@@ -501,6 +567,9 @@ def copy(self):
         clipboard.setText(contents)
 
 
+
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: Remove this blank line

File: spyder/widgets/variableexplorer/dataframeeditor.py [hunk 7]
```diff
@@ -542,6 +611,7 @@ def setup_and_check(self, data, title=''):
         self.resize(600, 500)
 
         self.dataModel = DataFrameModel(data, parent=self)
+        # self.dataTable = DataFrameFreezeView(self, self.dataModel)
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: Remove this comment

**Comment by ccordoba12** [comment]
@andfoy, thanks a lot for this. I did a first, quick review with style issues :-)

File: spyder/widgets/variableexplorer/dataframeeditor.py [hunk 8]
```diff
@@ -399,23 +401,102 @@ def reset(self):
...
+        self.setVerticalScrollMode(1)
+
+
 class DataFrameView(QTableView):
     """Data Frame view class"""
     def __init__(self, parent, model):
         QTableView.__init__(self, parent)
         self.setModel(model)
 
+        self.frozenTableView = FrozenTableView(self)
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: Please change this attribute to

    self.frozen_table_view

File: spyder/widgets/variableexplorer/dataframeeditor.py [hunk 9]
```diff
@@ -399,23 +401,102 @@ def reset(self):
...
+                       (self.frozenTableView.columnWidth(0) + 
+                        self.frozenTableView.columnWidth(1))
+            self.horizontalScrollBar().setValue(newValue)
+        return current
+
+    def scrollTo(self, index, hint):
+        if index.column() > 1:
+            QTableView.scrollTo(self, index, hint)
+
+    def updateFrozenTableGeometry(self):
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: We can move this to be a method of the `frozenTableView` class

File: spyder/widgets/variableexplorer/dataframeeditor.py [hunk 10]
```diff
@@ -399,23 +401,102 @@ def reset(self):
...
+    def __init__(self, parent):
+        QTableView.__init__(self, parent)
+        self.setModel(parent.model())
+        self.setFocusPolicy(Qt.NoFocus)
+        self.verticalHeader().hide()
+        if PYQT5:
+            self.horizontalHeader(). \
+            setSectionResizeMode(QHeaderView.Fixed)
+        else:
+            self.horizontalHeader(). \
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: Let's not add continuation lines to these statements (even if they are longer than 79 columns)

File: spyder/widgets/variableexplorer/dataframeeditor.py [hunk 11]
```diff
@@ -399,23 +401,102 @@ def reset(self):
...
         self.horizontalScrollBar().valueChanged.connect(
-                            lambda val: self.load_more_data(val, columns=True))
+                        lambda val: self.load_more_data(val, columns=True))
+        self.verticalScrollBar().valueChanged.connect(
+                        lambda val: self.load_more_data(val, rows=True))
         self.verticalScrollBar().valueChanged.connect(
-                               lambda val: self.load_more_data(val, rows=True))
+            self.frozenTableView.verticalScrollBar().setValue)
     
+    def updateSectionWidth(self, logicalIndex, oldSize, newSize):
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: Please follow pep8 here too: `def update_section_width`

File: spyder/widgets/variableexplorer/dataframeeditor.py [hunk 12]
```diff
@@ -399,23 +401,102 @@ def reset(self):
...
+        if logicalIndex == 0:
+            self.frozenTableView.setColumnWidth(0, newSize)
+            self.updateFrozenTableGeometry()
+
+    def updateSectionHeight(logicalIndex, oldSize, newSize):
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: Pep8

File: spyder/widgets/variableexplorer/dataframeeditor.py (hunk 9 above)
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: And also rename this method following pep8

File: spyder/widgets/variableexplorer/dataframeeditor.py [hunk 13]
```diff
@@ -399,23 +401,102 @@ def reset(self):
...
+        self.updateFrozenTableGeometry()
+
+    def moveCursor(self, cursorAction, modifiers):
+        current = QTableView.moveCursor(self, cursorAction, modifiers)
+        if cursorAction == self.MoveLeft and \
+                           current.column() > 1 and \
+                           self.visualRect(current).topLeft().x() < \
+                           (self.frozenTableView.columnWidth(0) + 
+                            self.frozenTableView.columnWidth(1)):
+            newValue = self.horizontalScrollBar().value() + \
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: pep8 for this variable, and there's no need of continuation lines if you enclose this whole expression in parenthesis

File: spyder/widgets/variableexplorer/dataframeeditor.py [hunk 14]
```diff
@@ -399,23 +401,102 @@ def reset(self):
...
+
+    def resizeEvent(self, event):
+        QTableView.resizeEvent(self, event)
+        self.updateFrozenTableGeometry()
+
+    def moveCursor(self, cursorAction, modifiers):
+        current = QTableView.moveCursor(self, cursorAction, modifiers)
+        if cursorAction == self.MoveLeft and \
+                           current.column() > 1 and \
+                           self.visualRect(current).topLeft().x() < \
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: Please move this computation to a variable

File: spyder/widgets/variableexplorer/dataframeeditor.py [hunk 15]
```diff
@@ -399,23 +401,102 @@ def reset(self):
...
+    def updateSectionHeight(logicalIndex, oldSize, newSize):
+        self.frozenTableView.setRowHeight(logicalIndex, newSize)
+
+    def resizeEvent(self, event):
+        QTableView.resizeEvent(self, event)
+        self.updateFrozenTableGeometry()
+
+    def moveCursor(self, cursorAction, modifiers):
+        current = QTableView.moveCursor(self, cursorAction, modifiers)
+        if cursorAction == self.MoveLeft and \
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: Also pep8 here, and avoid using the continuation operator (`\`)

File: spyder/widgets/variableexplorer/dataframeeditor.py (hunk 9 above)
Comment (andfoy) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: Well, the geometry update must be done by the parent table that contains the frozen column. Then this method should be here.

File: spyder/widgets/variableexplorer/dataframeeditor.py (hunk 11 above)
Comment (andfoy) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: Isn't this an Qt method override?

File: spyder/widgets/variableexplorer/dataframeeditor.py (hunk 12 above)
Comment (andfoy) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: See above comment

File: spyder/widgets/variableexplorer/dataframeeditor.py (hunk 15 above)
Comment (andfoy) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: Isn't this an Qt method override?

File: spyder/widgets/explorer.py [hunk 16]
```diff
@@ -562,6 +562,7 @@ def convert_notebook(self, fname):
         """Convert an IPython notebook to a Python script in editor"""
         try: 
             script = nbexporter().from_filename(fname)[0]
+            # print(script)
```
Comment (ccordoba12) [inline spyder/widgets/explorer.py]: Please remove this comment

File: spyder/widgets/variableexplorer/dataframeeditor.py [hunk 17]
```diff
@@ -15,13 +15,15 @@
 
 # Third party imports
 from pandas import DataFrame, Series
+from qtpy import PYQT5
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: Move this import to the line below, like this

    from qtpy import API, PYQT5

File: spyder/widgets/variableexplorer/dataframeeditor.py [hunk 18]
```diff
@@ -399,23 +401,103 @@ def reset(self):
         self.endResetModel()
 
 
+class FrozenTableView(QTableView):
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: Please add a docstring to this class to explain why it's needed and where you took it from (i.e. the Qt page I referred you to)

File: spyder/widgets/variableexplorer/dataframeeditor.py [hunk 19]
```diff
@@ -399,23 +401,103 @@ def reset(self):
...
         self.horizontalScrollBar().valueChanged.connect(
-                            lambda val: self.load_more_data(val, columns=True))
+                        lambda val: self.load_more_data(val, columns=True))
+        self.verticalScrollBar().valueChanged.connect(
+                        lambda val: self.load_more_data(val, rows=True))
         self.verticalScrollBar().valueChanged.connect(
-                               lambda val: self.load_more_data(val, rows=True))
+            self.frozen_table_view.verticalScrollBar().setValue)
     
+    def update_section_width(self, logical_index, old_size, new_size):
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: docstring

File: spyder/widgets/variableexplorer/dataframeeditor.py [hunk 20]
```diff
@@ -399,23 +401,103 @@ def reset(self):
...
+        if logical_index == 0:
+            self.frozen_table_view.setColumnWidth(0, new_size)
+            self.frozen_table_view.update_frozen_table_geometry()
+
+    def update_section_height(logical_index, old_size, new_size):
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: docstring

File: spyder/widgets/variableexplorer/dataframeeditor.py [hunk 21]
```diff
@@ -399,23 +401,103 @@ def reset(self):
...
+        overflow = self.MoveLeft and current.column() > 1
+        overflow = overflow and x_top_l < col_width
+
+        if cursor_action == overflow:
+            new_value = (self.horizontalScrollBar().value() + 
+                         x_top_l - col_width)
+            self.horizontalScrollBar().setValue(new_value)
+        return current
+
+    def scrollTo(self, index, hint):
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: docstring

**Comment by ccordoba12** [comment]
Thanks @andfoy for the improvements! I asked you to add docstrings to all methods and classes you introduced, so it's easier to understand what's going on with this code later.

Please also verify that your changes don't break the DataFrame editor when you have more than 100.000 rows and specially more than 60 columns.

**Comment by ccordoba12** [comment]
Also, please merge with 3.x to fix the style errors in CircleCI

**Comment by andfoy** [comment]
@ccordoba12 I tested against a dataframe of size 100000x70, and there were no errors or failures at all.

File: spyder/widgets/variableexplorer/dataframeeditor.py [hunk 22]
```diff
@@ -399,23 +400,134 @@ def reset(self):
         self.endResetModel()
 
 
+class FrozenTableView(QTableView):
+    """
+    This class implements a table with its first column frozen
+    For more information, visit:
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: Please change this line to

    For more information please see:

File: spyder/widgets/variableexplorer/dataframeeditor.py [hunk 23]
```diff
@@ -399,23 +400,134 @@ def reset(self):
...
+        self.setColumnWidth(0, parent.columnWidth(0))
+        self.setHorizontalScrollBarPolicy(Qt.ScrollBarAlwaysOff)
+        self.setVerticalScrollBarPolicy(Qt.ScrollBarAlwaysOff)
+        self.show()
+
+        self.setVerticalScrollMode(1)
+
+    def update_frozen_table_geometry(self):
+        """
+        Updates the frozen column size when an update occurs
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: Updates -> Update

File: spyder/widgets/variableexplorer/dataframeeditor.py [hunk 24]
```diff
@@ -399,23 +400,134 @@ def reset(self):
...
+                        lambda val: self.load_more_data(val, columns=True))
+        self.verticalScrollBar().valueChanged.connect(
+                        lambda val: self.load_more_data(val, rows=True))
         self.verticalScrollBar().valueChanged.connect(
-                               lambda val: self.load_more_data(val, rows=True))
+            self.frozen_table_view.verticalScrollBar().setValue)
     
+    def update_section_width(self, logical_index, old_size, new_size):
+        """
+        Updates the horizontal width of the frozen column when a
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: Updates -> Upddate

File: spyder/widgets/variableexplorer/dataframeeditor.py [hunk 25]
```diff
@@ -399,23 +400,134 @@ def reset(self):
...
+        change takes place in the first column of the table 
+        """
+        if logical_index == 0:
+            self.frozen_table_view.setColumnWidth(0, new_size)
+            self.frozen_table_view.update_frozen_table_geometry()
+
+    def update_section_height(logical_index, old_size, new_size):
+        """
+        Updates the vertical width of the frozen column when a
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: Updates -> Update

File: spyder/widgets/variableexplorer/dataframeeditor.py [hunk 26]
```diff
@@ -399,23 +400,134 @@ def reset(self):
...
+        """
+        Updates the frozen column dimensions when the enclosing 
+        window of this table reports a dimension change 
+        """
+        QTableView.resizeEvent(self, event)
+        self.frozen_table_view.update_frozen_table_geometry()
+
+    def moveCursor(self, cursor_action, modifiers):
+        """
+        Updates the table position, along with the frozen column
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: Updates -> Update

File: spyder/widgets/variableexplorer/dataframeeditor.py [hunk 27]
```diff
@@ -399,23 +400,134 @@ def reset(self):
...
+            self.setColumnHidden(col, True)
+
+        self.setColumnWidth(0, parent.columnWidth(0))
+        self.setHorizontalScrollBarPolicy(Qt.ScrollBarAlwaysOff)
+        self.setVerticalScrollBarPolicy(Qt.ScrollBarAlwaysOff)
+        self.show()
+
+        self.setVerticalScrollMode(1)
+
+    def update_frozen_table_geometry(self):
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: Let's change the name of this method to `update_geometry`

File: spyder/widgets/variableexplorer/dataframeeditor.py [hunk 28]
```diff
@@ -399,23 +400,134 @@ def reset(self):
...
+    def moveCursor(self, cursor_action, modifiers):
+        """
+        Updates the table position, along with the frozen column
+        when the cursor (selector) inside, changes its position  
+        """
+        current = QTableView.moveCursor(self, cursor_action, modifiers)
+        
+        col_width = (self.frozen_table_view.columnWidth(0) + 
+                     self.frozen_table_view.columnWidth(1))
+        x_top_l = self.visualRect(current).topLeft().x()
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: Please the name of this variable to `topleft_x`

**Comment by ccordoba12** [comment]
Last comments, and then this is definetly ready :-)

**Pull request merged**